planning/MartenV4.md

```diff
@@ -0,0 +1,405 @@
+
+### Event Store Partitioning or Sharding
+
+The event sourcing in Marten has limited scalability because of its dependence today upon appending events to 
+one table. Closely related to the storage is ways to segment the asynchronous projection support to parallelize
+the work across multiple threads or active nodes.
+
+Here are some ideas:
+
+* Allow the user to define multiple event logs in the database. This way you'd have multiple stream/event tables for completely
+  different kinds of streams
+* Postgresql native partitioning by tenant id where that's useful
+* Aggregation keys -- this is tightly coupled to the projections. If there's a projection that is done by something other than tenant id or stream id, try to persist the 
+  key for however the events/streams are grouped to stream and/or event storage. Thinking about data elements like "Region" or "Country" where you might be doing
+  aggregations across streams or tenants. Maybe that's useful?
+* Archive streams that reach some kind of termination event? Move everything to an archive table off to the side? Keeps the event log table cleaner
+* Artificial segmenting of the streams. Use some kind of algorithm that could see stream ids, and assign them to 2 or more "cohorts", then persist that "cohort key" to 
+  both the stream and event data. This would potentially help for the async daemon to parallelize work. It would definitely help out for "live" projections by having 
+  smaller tables to work through
+
+### Projections
+
+Thoughts:
+
+Functionality wise, the lifecycle of projections are:
+
+* "Live" projections built on demand from the raw events. This only changes for v4 to take advantage of new snapshotting capabilities
+* "Inline" projections built as part of the same unit of work as capturing the source events. See the section on *Inline Projections* because this
+  will change internally for v4
+* "Async" projections built up in a background process of some sort. I think this turns into a near re-write for v4
+
+By *type*, the projection types *could* be:
+
+* Aggregate document by stream (existing)
+* Aggregate document by tenant id (new)
+* Execute SQL statement by event type (new) -- so, insert/update/delete. This comes up a lot, and it should be low hanging fruit
+* Transform a single event to a document --> needs to become just "carry out some kind of change to a document" and probably ends up sharing more mechanics with the aggregation than it used to
+* Aggregate by arbitrary values of either the stream or event types -- but if we do this, I strongly vote to persist whatever the differentiation value is in the database to make projection rebuilds and the async daemon more "parallelizable"
+
+
+Other thoughts:
+
+* I want to completely throw away the current `Aggregator` implementations. It was just too naive for real world usage and the internals have gotten ugly
```

Comment:
Personally, I'd like to get rid of the different type of projections. Right now each of them has a separate implementation. While all of them so "live" - aggregations, "inline" - snapshots, "viewprojection" - document should be just different with the implemntation details and some optimisations as all of them are just different types of projections.
That was also the idea behind my PR https://github.com/JasperFx/marten/pull/1322. So unify at least on the interface plus the used internally components level the approach.

Reply:
I don't completely agree here. I don't think it's a big concern to have a unifying interface. Possibly a unified way to *define* the behavior of the projections, then rely a bit on runtime codegen to optimize the behavior

Reply:
Maybe unifying interface is an oversimplification. I mean more about using the same blocks of code and logic where possible. Current implementation even though they are conceptually similar stuff they have totally separated implementations with different conventions (also from the API perspective).

Also, I'd like to do the event transformations more usable. They're a nice tool for eg. creating an audit log. But their API doesn't allow to use event metadata plus it doesn't have useful helpers as ViewProjection. That seems to be low hanging fruit.

Reply:
I think we could expand the existing `ViewProjection` and incorporate the transformer class idea in such a way that we could cover every possible type of projection from events to documents.

The projection to flat tables may or may not fit into that.